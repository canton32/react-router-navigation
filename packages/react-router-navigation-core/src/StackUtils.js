/* @flow */

import React from 'react'
import { matchPath } from 'react-router'
import type { Location, MatchPathOptions } from 'react-router'
import type { Route } from './TypeDefinitions'

// Test if current stack item should be updated
export const shouldUpdate = (
  currentItem: MatchPathOptions,
  nextItem: MatchPathOptions,
  currentLocation: Location,
  nextLocation: Location,
): boolean => {
  // Get entries and matchs
  const matchCurrentRoute = matchPath(currentLocation.pathname, currentItem)
  const matchNextRoute = matchPath(nextLocation.pathname, nextItem)
  return (
    // Test if pathames are different
    currentLocation.pathname !== nextLocation.pathname &&
    // case 1) basic pathname
    (currentItem.path !== nextItem.path ||
      // case 2) pathname with query params
      // ex: with same path article/:id,
      //     pathname article/2 !== article/3
      (matchCurrentRoute !== null &&
        matchNextRoute !== null &&
        Object.keys(matchCurrentRoute.params).length !== 0 &&
        Object.keys(matchNextRoute.params).length !== 0 &&
        currentLocation.pathname !== nextLocation.pathname))
  )
}

// Get stack item from a specific route
export const get = <Item>(items: Array<Item>, route: Route): Item => ({
  ...route,
  ...items.find(item => {
    return item && item.key && route.routeName === item.key
  }),
})

// Generate unique key
export const createKey = (route: Route): string => {
  return `${route.key}@@${Math.random()
    .toString(10)
    .slice(1)}`
}

// Get current route from a specific history location
export const getRoute = (stack: Array<Object>, location: Location): ?Route => {
  const { pathname } = location
  const item = stack.find(stackItem => {
    return matchPath(pathname, stackItem)
  })
  if (!item || !item.key) return null
  return {
    key: createKey(item),
    routeName: item.key,
    match: matchPath(pathname, item),
  }
}

// Render a subview with props
export const renderSubView = (render: Function, additionalProps?: * = {}) => (
  ownProps: *,
): React$Element<*> => {
  const props = { ...additionalProps, ...ownProps }
  const {
    cards,
    tabs,
    scene,
    route,
    navigationState: { routes, index },
  } = props
  const item = {
    ...props,
    ...get(cards || tabs, (scene && scene.route) || route || routes[index]),
  }
  return render(
    Object.keys(item).reduce((acc, key) => {
      const value = item[key]
      if (value === undefined) return acc
      return { ...acc, [key]: value }
    }, {}),
    props,
  )
}

// Build stack with React elements
export const build = <Item>(
  children: Array<React$Element<Item>>,
  oldBuild?: Array<Item>,
): Array<Item> => {
  return React.Children.toArray(children).reduce((stack, child, index) => {
    const item = Object.keys(child.props).reduce((props, key) => {
      if (key === 'path') {
        return {
          ...props,
          key: child.props[key],
        }
      } else if (
        key === 'render' ||
        key === 'component' ||
        key === 'children'
      ) {
        return {
          ...props,
          [key]:
            oldBuild && oldBuild[index] // $FlowFixMe
              ? oldBuild[index][key]
              : () => React.cloneElement(child),
        }
      }
      return props
    }, child.props)
    if (!item.key) return stack
    return [...stack, item]
  }, [])
}
