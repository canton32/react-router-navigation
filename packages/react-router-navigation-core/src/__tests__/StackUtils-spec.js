import React from 'react'
import { Text } from 'react-native'
import { Router, Route } from 'react-router'
import createHistory from 'history/createMemoryHistory'
import renderer from 'react-test-renderer'
import * as StackUtils from './../StackUtils'

const { build, shouldUpdate, get, getRoute, createKey } = StackUtils

describe('build util', () => {
  it('works correctly', () => {
    const children = [
      <Route
        path="hey"
        render={function() {
          return null
        }}
      />,
      <Route
        path="hello"
        component={function() {
          return null
        }}
        title="Hello"
      />,
    ]
    const results = [
      {
        path: 'hey',
        key: 'hey',
      },
      {
        path: 'hello',
        key: 'hello',
        title: 'Hello',
      },
    ]
    expect(build(children)).toMatchObject(results)
    expect(typeof build(children)[0].render).toBe('function')
    expect(typeof build(children)[1].component).toBe('function')
  })

  it('creates correctly render method', () => {
    const children = [
      <Route
        path="/"
        render={({ location }) => <Text>{location.pathname}</Text>}
      />,
    ]
    const stack = build(children)
    const history = createHistory()
    const component = renderer.create(
      <Router history={history}>{stack[0].render()}</Router>,
    )
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})

describe('shouldUpdate util', () => {
  it('returns true when paths are different', () => {
    const currentCard = { path: '/foo' }
    const nextCard = { path: '/bar' }
    const currentRouterHistory = createHistory({
      initialEntries: ['/foo'],
    })
    const nextRouterHistory = createHistory({
      initialIndex: 1,
      initialEntries: ['/foo', '/bar'],
    })
    const input = shouldUpdate(
      currentCard,
      nextCard,
      currentRouterHistory.location,
      nextRouterHistory.location,
    )
    expect(input).toBeTruthy()
  })
  it('returns false when paths are equal', () => {
    const currentCard = { path: '/foo' }
    const nextCard = { path: '/foo' }
    const currentRouterHistory = createHistory({
      initialEntries: ['/foo'],
    })
    const nextRouterHistory = createHistory({
      initialIndex: 0,
      initialEntries: ['/foo'],
    })
    const input = shouldUpdate(
      currentCard,
      nextCard,
      currentRouterHistory.location,
      nextRouterHistory.location,
    )
    expect(input).toBeFalsy()
  })
  it('returns true when paths (with params) are equal and pathnames are different (1)', () => {
    const currentCard = { path: '/article/:id' }
    const nextCard = { path: '/article/:id' }
    const currentRouterHistory = createHistory({
      initialIndex: 1,
      initialEntries: ['/article/1', '/article/2'],
    })
    const nextRouterHistory = createHistory({
      initialEntries: ['/article/1', '/article/2'],
    })
    const input = shouldUpdate(
      currentCard,
      nextCard,
      currentRouterHistory.location,
      nextRouterHistory.location,
    )
    expect(input).toBeTruthy()
  })
  it('returns true when paths (with params) are equal and pathnames are different (2)', () => {
    const currentCard = { path: '/article/:id' }
    const nextCard = { path: '/article/:id' }
    const currentRouterHistory = createHistory({
      initialEntries: ['/article/1'],
    })
    const nextRouterHistory = createHistory({
      initialIndex: 1,
      initialEntries: ['/article/1', '/article/2'],
    })
    const input = shouldUpdate(
      currentCard,
      nextCard,
      currentRouterHistory.location,
      nextRouterHistory.location,
    )
    expect(input).toBeTruthy()
  })
  it('returns false when paths (without params) are equal and pathname are different', () => {
    const currentCard = { path: '/app' }
    const nextCard = { path: '/app' }
    const currentRouterHistory = createHistory({
      initialEntries: ['/app'],
    })
    const nextRouterHistory = createHistory({
      initialIndex: 1,
      initialEntries: ['/app', '/app/article'],
    })
    const input = shouldUpdate(
      currentCard,
      nextCard,
      currentRouterHistory.location,
      nextRouterHistory.location,
    )
    expect(input).toBeFalsy()
  })
})

describe('get util', () => {
  it('works correctly', () => {
    const route = { key: '/index@@h9208990', routeName: '/index' }
    const currentCard = {
      key: '/index',
      routeName: '/index',
      title: 'Index',
    }
    const cards = [
      {
        key: '/',
        title: 'Root',
      },
      currentCard,
    ]
    expect(get(cards, route)).toEqual(currentCard)
  })
})

describe('getRoute util', () => {
  it('works correctly', () => {
    const stack = [{ key: '/foo', path: '/foo' }, { key: '/bar', path: '/bar' }]
    const history = createHistory({ initialEntries: ['/bar'] })
    expect(getRoute(stack, history.location).routeName).toEqual('/bar')
  })
})

describe('createKey util', () => {
  it('createKey() works correctly', () => {
    const key = createKey({ key: '/foo' })
    expect(key.slice(0, 4)).toEqual('/foo')
  })
})
