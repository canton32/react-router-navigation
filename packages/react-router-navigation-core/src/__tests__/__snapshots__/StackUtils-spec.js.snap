// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`build util creates correctly render method 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
>
  /
</Text>
`;
